store/store.go

```diff
@@ -326,13 +321,20 @@ func (bs *BlockStore) PruneBlocks(height int64) (uint64, error) {
 	return pruned, nil
 }
 
-// SaveBlock persists the given block, blockParts, and seenCommit to the underlying db.
+// SaveBlock persists the given block, blockParts, and seenCommit to the
+// underlying db, while also putting the block data onto IPFS
 // blockParts: Must be parts of the block
-// seenCommit: The +2/3 precommits that were seen which committed at height.
+// seenCommit: The +2/3 precommits that were seen which
+// committed at height.
 //             If all the nodes restart after committing a block,
 //             we need this to reload the precommits to catch-up nodes to the
 //             most recent height.  Otherwise they'd stall at H-1.
-func (bs *BlockStore) SaveBlock(block *types.Block, blockParts *types.PartSet, seenCommit *types.Commit) {
+func (bs *BlockStore) SaveBlock(
```

Comment:
I am kinda concerned about this method as we have PutBlock already being called on consensus and I don't think relying just on Blockstore.SaveBlock instead for networking intensely operation is a good idea. In case we save Block two times(overheads) just to persist Meta, it sounds reasonable for me now to change blockstore into MetaStore or HeaderStore and no more handle block data read/writes at all.

Reply:
>  it sounds reasonable for me now to change blockstore into MetaStore or HeaderStore and no more handle block data read/writes at all.

You are right about this. But we currently already store the block twice. We need to change this indeed.
The question is if in this PR or in a separate one. Swapping the blockstore with the header store is basically reviving and slightly rethinking @marbar3778's PR: https://github.com/lazyledger/lazyledger-core/pull/218. 

My suggestion was to tackle this in a separate PR but if you feel strongly about this and @evan-forbes agrees, I'm OK with doing this in this PR as well.

Reply:
This actually might require more thoughts. Also the API @renaynay started which will enable retrieving the state-relevant portion might be necessary for this.
See: https://github.com/lazyledger/lazyledger-core/pull/218#issuecomment-803381445

Reply:
But to not be blocked on that API (https://github.com/lazyledger/lazyledger-core/issues/221), we could make it that wherever the Txs are needed for replay, the block is downloaded via IPFS if not locally available anyways.

Reply:
Yeah, Ismail also shared your concern over the this, particularly when loading block data. I think it's improved slightly by using the offline version of the dag api https://github.com/lazyledger/lazyledger-core/pull/374/commits/8619f275bd635f9d95891d755d3e9b700de514d6 . Unfortunately, until we add some caching mechanism I don't think we have another way to ensure that the data in the block being saved is added to IPFS other than `PutBlock`. 

Out of curiosity, is saving data to IPFS that network intensive? I know there's certainly overhead, as opposed to saving to local only DB, but unlike fetching data there's no blocking involved, right? 

I _really_ like the idea of breaking up the `Blockstore` struct to two different entities, one that strictly uses the local store, and another focused on retrieving block data from IPFS. I think that makes a lot of sense for LazyLedger. However, I do feel that is out of the scope of this PR. This PR's main purpose is to get us to the next step of removing `BlockID` and stop saving block data twice.

Reply:
I am ok to tackle this in a separate PR.

> Out of curiosity, is saving data to IPFS that network-intensive? 

It is, as down the road Bitswap provides data to DHT and that's kinda network-intensive. Further, after https://github.com/lazyledger/lazyledger-core/pull/375 is merged, PutBlock may even become blocking for providing. The providing itself is also a concern as it definitely should be out of Blockstore and currently there is no way to disable it.

Reply:
Again, I am fully ok to work over this in another PR, as a proper solution for this is not obvious and would probably require some level of refactoring over multiple components

Reply:
Yeah, let's try to tackle the story behind this in another PR. That PR will require some clearly outlined reasoning and an implementation plan in an issue or maybe even (as part of) a full-blown ADR (specifically around storage, i.e. storage for validator nodes, storage for partial nodes that passively participate in the consensus–those that will replace the current tendermint fullnodes). 

I think on master, only block-proposers call PutBlock. The other nodes currently only explicitly store the block data via the tendermint store (#218 tries to change that). I think for this PR it's OK to skip the online/offline story. 
BTW, I could imagine that if PutBlock is called twice and there is no update on the data, the network IO would not really be triggered (not sure though, would need to look into the interplay between the ipfs block store and bitswap).